synapse/config/workers.py

```diff
@@ -183,10 +191,10 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
         )
 
         # A map from instance name to host/port of their HTTP replication endpoint.
-        instance_map = config.get("instance_map") or {}
-        self.instance_map = {
-            name: InstanceLocationConfig(**c) for name, c in instance_map.items()
-        }
+        # instance_map = config.get("instance_map") or {}
+        self.instance_map = parse_obj_as(
+            Dict[str, InstanceLocationConfig], config.get("instance_map") or {}
```

Comment:
Prefer `config.get("instance_map", {})` here; As written above, this would allow `instance_map` to be any false type that can be parsed from yaml (`0`, `False`, `""`, `[]`, `{}`, `None`.

Additionally: if the validation fails this will raise a ValidationError. We would like it to raise a ConfigError instead. Some kind of helper function, like a simpler version of

https://github.com/matrix-org/synapse/blob/68a671731207645f693e4e48676781b9a1acb838/synapse/http/servlet.py#L781-L807

in synapse.config._util would likely be useful here and in the future.

Reply:
I gave this a try in [09657ff](https://github.com/matrix-org/synapse/pull/15431/commits/09657ff56c0f095b5a54f386cc04a24e22902537). I'm not convinced I did it right although it passes every test I know to throw at it.
* `TypeAlias` in particular feels like a 'Get-out-of-jail-free' Card. Mypy wouldn't let me pass a `Type[Model]` in, as `parse_obj_as()` wouldn't except it. Possibly could just go back to a `for` loop and use `parse_obj()`?
* The return annotation being the only place to utilize `Model` feels wrong somehow. In the interest of making it abstract, I didn't think using `InstanceLocationConfig` directly was appropriate either. It's possible(although unlikely) this could be used for `stream_writers` in the future. I know of no other 'Map' style configuration done in the `WorkersConfig` section where this would be appropriate.

Reply:
Looks like `TypeAlias` isn't available on Python 3.7, so I guess will have to think of something else there.